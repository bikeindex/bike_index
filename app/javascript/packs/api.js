/*
  Private
*/

const url = urn => [process.env.BASE_URL, urn].join("/");

const serialSearchUrl = serial =>
  url(`api/v2/bikes_search?serial=${serial}`);

const fuzzySearchUrl = serial =>
  url(`api/v2/bikes_search/close_serials?serial=${serial}`);

const partialMatchSerialSearchUrl = params => {
  const query = queryString(params)
  return url(`api/v3/search/serials_containing?${query}`);
}

const serialCloseSearchUrl = params => {
  const query = queryString(params)
  return url(`api/v3/search/close_serials?${query}`);
}

const serialExternalSearchUrl = ({ serial }) => {
  const query = queryString({ serial })
  return url(`api/v3/search/external_registries?${query}`);
}

const request = async url => {
  const resp = await fetch(url);
  const json = await resp.json();
  return json;
};

const queryString = ({ serial, stolenness, location, query, query_items }) => {
  const params = {};
  if (serial) { params.serial = serial; }
  if (stolenness) { params.stolenness = stolenness; }
  if (location) { params.location = location; }
  if (query) { params.query = query; }
  if (query_items) { params.query_items = query_items; }

  return Object.keys(params).map(k => `${k}=${params[k]}`).join("&");
}

/*
  Public
*/

const fetchSerialResults = serial => {
  const url = serialSearchUrl(serial);
  return request(url);
};

const fetchFuzzyResults = serial => {
  const url = fuzzySearchUrl(serial);
  return request(url);
};

const fetchPartialMatchSearch = interpretedParams => {
  const url = partialMatchSerialSearchUrl(interpretedParams);
  return request(url);
}

const fetchSerialCloseSearch = interpretedParams => {
  const url = serialCloseSearchUrl(interpretedParams);
  return request(url);
}

const fetchSerialExternalSearch = ({ raw_serial }) => {
  const url = serialExternalSearchUrl({ serial: raw_serial });
  return request(url);
}

export default {
  fetchSerialResults,
  fetchFuzzyResults,
  fetchSerialExternalSearch,
  fetchSerialCloseSearch,
  fetchPartialMatchSearch,
};
