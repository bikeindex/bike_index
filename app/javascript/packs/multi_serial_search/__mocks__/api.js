const fetchSerialResults = async () => new Promise(resolve => resolve({}));

export {
  fetchSerialResults,
};
