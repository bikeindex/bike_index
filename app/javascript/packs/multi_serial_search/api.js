/* TODO: put some of the private implementation into a utility helper */

/*
  Private
*/

const isDevelopment = process.env.RAILS_ENV === 'development';

const serialSearchUrl = serial => {
  const urn = `api/v1/bikes?multi_serial_search=true&serial=${serial}`;
  return isDevelopment
    ? `http://localhost:3001/${urn}`
    : `https://bikeindex.org/${urn}`;
};

const request = async url => {
  const resp = await fetch(url);
  const json = await resp.json();
  return json;
};

/*
  Public
*/

const fetchSerialResults = async serial => {
  const url = serialSearchUrl(serial);
  const results = await request(url);
  return results;
};

export { fetchSerialResults };
