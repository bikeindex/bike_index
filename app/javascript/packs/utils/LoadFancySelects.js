import selectize from "selectize";

// Right now we're using selectize.js (https://selectize.github.io/selectize.js/) to make our select boxes fancy.
// This could be changed to some other library at some point - this makes it possible for us to abstract that away.
// Just add the `unfancy` and `fancy-select` classes to a select box and it will be a fancy select box!
const LoadFancySelects = () => {
  // Gross, check if
  $(".unfancy.fancy-select.no-restore-on-backspace select").selectize({
    create: false,
    plugins: []
  });

  $(".unfancy.fancy-select select").selectize({
    create: false,
    plugins: ["restore_on_backspace"]
  });
  // Remove them so we don't initialize twice
  $(".unfancy.fancy-select, .unfancy.fancy-select-placeholder").removeClass(
    "unfancy"
  );
};

export default LoadFancySelects;
