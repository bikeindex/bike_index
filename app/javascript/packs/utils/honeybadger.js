import Honeybadger from "honeybadger-js";
import log from "./log";

const environment = process.env.RAILS_ENV;

/* webpacker can't access this exported env */
const apiKey =
  environment === "production" ? process.env.HONEYBADGER_FRONTEND_API_KEY : "foo";

const config = {
  apiKey,
  environment: process.env.RAILS_ENV,
  disabled: environment !== "production"
};
const honeybadger = Honeybadger.configure(config);

Honeybadger.beforeNotify(notice => {
  if (environment === "development") {
    return log.debug("Error", notice);
  }
});
export default honeybadger;
