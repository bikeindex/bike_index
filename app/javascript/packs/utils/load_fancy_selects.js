import selectize from "selectize";

// Right now we're using selectize.js (https://selectize.github.io/selectize.js/) to make our select boxes fancy.
// This could be changed to some other library at some point - this makes it possible for us to abstract that away.
// Just add the `unfancy` and `fancy-select` classes to a select box and it will be a fancy select box!
const LoadFancySelects = () => {
  $(".unfancy.fancy-select.no-restore-on-backspace select").selectize({
    create: false,
    plugins: []
  });
  // Remove, because we've already added them
  $(".unfancy.fancy-select.no-restore-on-backspace").removeClass("unfancy");

  $(".unfancy.fancy-select select").selectize({
    create: false,
    plugins: ["restore_on_backspace"]
  });
  // Remove them so we don't initialize twice
  $(".unfancy.fancy-select").removeClass("unfancy");
};

export default LoadFancySelects;
