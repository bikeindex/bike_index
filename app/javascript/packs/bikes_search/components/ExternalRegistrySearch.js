/* eslint import/no-unresolved: 0 */

import React from "react";

import BikeSearch from "./BikeSearch";
import ExternalRegistrySearchResult from "./ExternalRegistrySearchResult";
import { fetchSerialExternalSearch } from "../../api";

class ExternalRegistrySearch extends BikeSearch {
  componentName = "ExternalRegistrySearch";
  fetchBikes = fetchSerialExternalSearch;
  headerDomId = "js-external-registry-search-header";
  resultComponent = ExternalRegistrySearchResult;

  componentWillMount() {
    const { stolenness, query, serial } = this.props.interpretedParams;
    // Only search external registries if looking for
    // stolen/all bikes with no query
    if (stolenness === "non" || query || !serial) { return; }
    super.componentWillMount();
  }
};

export default ExternalRegistrySearch;
