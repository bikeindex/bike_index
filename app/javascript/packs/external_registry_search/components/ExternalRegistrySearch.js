/* eslint import/no-unresolved: 0 */

import React, { Fragment, Component } from "react";

import ExternalRegistrySearchResult from "./ExternalRegistrySearchResult";
import { fetchSerialExternalSearch } from "../../api";
import Loading from "../../Loading";
import honeybadger from "../../utils/honeybadger";

class ExternalRegistrySearch extends Component {
  state = {
    loading: false,
    results: []
  }

  componentWillMount() {
    const { stolenness, query, serial } = this.props;
    // Only search external registries if looking for
    // stolen/all bikes with no query
    if (stolenness === "non" || query) { return; }

    this.resultsBeingFetched();

    fetchSerialExternalSearch(serial)
      .then(this.resultsFetched)
      .catch(this.handleError);
  }

  resultsBeingFetched = () => {
    this.setState({ loading: true });
    this.toggleHeader({ isLoading: true });
  }

  resultsFetched = ({ bikes, error }) => {
    this.setState({ results: bikes || [], loading: false });
    this.toggleHeader({ isLoading: false, resultsCount: this.state.results.length });
    if (error) { this.handleError(error) }
  }

  handleError = error => {
    honeybadger.notify(error, { component: "MultiSerialSearch" });
  }

  toggleHeader = ({ isLoading, resultsCount }) => {
    const header = document.getElementById("js-external-registry-search-header");
    if (!header) { return; }

    header.childNodes.forEach(node => node.classList && node.classList.add("d-none"));

    const titleDisplay = (isLoading)
          ? "loading"
          : (resultsCount)
          ? "loaded"
          : "loaded-none";

    const sectionTitle = header.getElementsByClassName(titleDisplay)[0];
    sectionTitle.classList.remove("d-none");
  }

  render() {
    if (this.state.loading) {
      return <Loading />;
    }

    return (
      <Fragment>
        <ul className="bike-boxes">
          {this.state.results.map(bike => <ExternalRegistrySearchResult key={bike.id} bike={bike}/>)}
        </ul>
      </Fragment>
    );
  }
};

export default ExternalRegistrySearch;
