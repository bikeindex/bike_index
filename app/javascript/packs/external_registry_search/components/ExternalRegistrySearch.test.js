/* eslint import/no-unresolved: 0 */

import React from "react";
import { shallow, mount } from "enzyme";
import toJson from "enzyme-to-json";

import { event } from "helpers/utils";
import { externalBikeResponse } from "helpers/bikes";

import * as api from "../../api";
import ExternalRegistrySearch from "./ExternalRegistrySearch";

describe("<ExternalRegistrySearch />", () => {
  it("matches the snapshot when loading", () => {
    const tree = shallow(<ExternalRegistrySearch serial="23545"/>);
    expect(toJson(tree)).toMatchSnapshot();
  });

  it("has a loading indicator when loading", () => {
    const wrap = mount(<ExternalRegistrySearch serial="12345"/>);
    expect(wrap.exists("Loading")).toBe(true);
  });

  it("is not loading after fetching results", () => {
    const wrap = mount(<ExternalRegistrySearch serial="12345"/>);
    wrap.setState({ loading: false, bikes: externalBikeResponse });
    expect(wrap.exists("Loading")).toBe(false);
  });
});
