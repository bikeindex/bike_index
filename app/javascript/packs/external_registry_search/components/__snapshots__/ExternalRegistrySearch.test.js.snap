// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ExternalRegistrySearch /> matches the snapshot when loading 1`] = `<Loading />`;
